Importation from Maintenance Branch

SNBeast opened the pull request:
- [x] Test
- [x] Test better by having a GUI to use

SNBeast committed:
Gen I - Part 1

SNBeast committed:
Fix bag indexing

SNBeast committed:
Gen I usable and levels fixed

SNBeast committed:
Gen I fixes and PK1 names

SNBeast committed:
Most of Gen II support

Along with some Gen I fixes
Some edge cases left to fix

SNBeast committed:
Gen II fixes for i18n and items

SNBeast committed:
Gen I and II forward transfers

SNBeast committed:
Back transfers to Gen I and II

SNBeast committed:
Gen I and II species nickname detection

(forgot to put in last commit)

SNBeast committed:
Remove warnings, pseudo-stubs, and fix back-transfer bug

piepie62 commented on include/utils/utils.hpp:
```diff
@@ -100,6 +101,18 @@ namespace StringUtils
+    void setString2(u8* data, const std::string_view& v, int ofs, int len, pksm::Language lang, int padTo = 0,
+        u8 padWith = 0x50);
+    
+    std::string getTradeOT(pksm::Language lang);
```
Why is this in the StringUtils namespace?

SNBeast replied:
It's exclusively used by functions in that namespace, and is a utility function that returns a string.

Though with the next commit it'll be used for a default OT for language-incompatible imports.

piepie62 commented on source/pkx/PK1.cpp:
```diff
@@ -0,0 +1,630 @@
+        pk2->egg(false);
+        pk2->otFriendship(pk2->baseFriendship());
+        pk2->language(language());
+        pk2->statExperience(Stat::HP, ev(Stat::HP));
```
These are probably supposed to be pk2->statExperience(<stat>, statExperience(<stat>)). That said, the "statExperience" model will be discussed in the Discord first; I'm not sure whether or not I like the way this is handled.

piepie62 commented on source/pkx/PK1.cpp:
```diff
@@ -0,0 +1,630 @@
+            case Species::Koffing:
+            case Species::Weezing:
+            case Species::Mew:
+                pk7->abilityNumber(1);
```
These two should be merged into setAbility(0)

piepie62 commented on source/pkx/PK1.cpp:
```diff
@@ -0,0 +1,630 @@
+                pk7->ability(pk7->abilities(0));
+                break;
+            default:
+                pk7->abilityNumber(4);
```
These two should be merged into setAbility(2)

piepie62 commented on source/pkx/PK1.cpp:
```diff
@@ -0,0 +1,630 @@
+    }
+    void PK1::hpType(Type v)
+    {
+        // TODO
```
Get this done, please

piepie62 commented on source/pkx/PK1.cpp:
```diff
@@ -0,0 +1,630 @@
+            case 0:
+                return;
+            default:
+                // TODO
```
Get this done, please

piepie62 commented on source/pkx/PK1.cpp:
```diff
@@ -0,0 +1,630 @@
+    }
+    Nature PK1::nature() const
+    {
+        return Nature{u8(experience() % 25)};
```
*Really*? That seems... terrifying

SNBeast replied:
Yep. From Bulbapedia:
> A transferred Pokémon's nature is that nature whose index number corresponds to the Pokémon's experience modulo 25.

piepie62 commented on source/pkx/PK1.cpp:
```diff
@@ -0,0 +1,630 @@
+    {
+        u8 i      = 1;
+        u8 xpType = expType();
+        while (experience() >= expTable(i, xpType) && ++i < 100)
```
Isn't this the generation where you can have level 101+ Pokémon? Should we handle that if so?

SNBeast replied:
Resolved elsewhere: decided against

piepie62 commented on source/pkx/PK1.cpp:
```diff
@@ -0,0 +1,630 @@
+    {
+        if (isParty()) shiftedData[33] = v;
+    }
+    int PK1::boxLevel() const
```
Not sure I like this; it should probably be directly accessed via `level()` with that secondarily setting experience. Feel free to convince me otherwise

SNBeast replied:
I don't agree, as this value doesn't update while in the party afaik.

piepie62 commented on source/pkx/PK2.cpp:
```diff
@@ -0,0 +1,735 @@
+        pk1->egg(false);
+        pk1->otFriendship(70);
+        pk1->language(language());
+        pk1->statExperience(Stat::HP, ev(Stat::HP));
```
Same as for PK1

piepie62 commented on source/pkx/PK2.cpp:
```diff
@@ -0,0 +1,735 @@
+    }
+    void PK2::alternativeForm(u16 v)
+    {
+        // TODO
```
Do this, please

piepie62 commented on source/pkx/PK2.cpp:
```diff
@@ -0,0 +1,735 @@
+    }
+    void PK2::hpType(Type v)
+    {
+        // TODO
```
Do this, please

piepie62 commented on source/pkx/PK2.cpp:
```diff
@@ -0,0 +1,735 @@
+    {
+        u8 i      = 1;
+        u8 xpType = expType();
+        while (experience() >= expTable(i, xpType) && ++i < 100)
```
Or was it G2 that could have level 100+? Both of them, perhaps?

SNBeast replied:
Resolved elsewhere: decided against

piepie62 commented on source/pkx/PK3.cpp:
```diff
@@ -452,6 +454,56 @@ namespace pksm
+        pk2->statExperience(Stat::SPD, ev(Stat::SPD) * ev(Stat::SPD));
+        pk2->statExperience(Stat::SPATK, ev(Stat::SPATK) * ev(Stat::SPATK));
+
+        pk2->move(0, move(0));
```
These need to be checked so that it can't transfer back a set of non-existent moves

piepie62 commented on source/pkx/PKX.cpp:
```diff
@@ -654,7 +647,42 @@ namespace pksm
+                                                        : PK2::INT_LENGTH_WITH_NAMES);
+                    break;
+                case Generation::THREE:
+                    boxlen = PK3::BOX_LENGTH;
```
These should all be PARTY_LENGTH; the point of this function is to clone the current Pokémon as a party Pokémon

piepie62 commented on source/sav/Sav.cpp:
```diff
@@ -84,6 +94,24 @@ namespace pksm
     }
+    std::unique_ptr<Sav> Sav::checkGBType(std::shared_ptr<u8[]> dt, size_t length)
+    {
+        u32 versionAndLanguage = Sav2::getVersion(dt);
```
Should be changed to follow API suggestion about this function

piepie62 commented on source/sav/Sav.cpp:
```diff
@@ -191,6 +219,10 @@ namespace pksm
         switch (generation())
         {
+            case Generation::ONE:
+                return pk.convertToG1(*this);
```
My understanding is that this doesn't handle cross-language transfer. Should it?

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+        if (lang == Language::ENG) lang = StringUtils::guessLanguage12(otName());
+    }
+
+    u32 Sav2::getVersion(std::shared_ptr<u8[]> dt)
```
This function should return an std::tuple<GameVersion, Language, bool>

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -0,0 +1,472 @@
+        OFS_BAG      = japanese ? 0x25C4 : 0x25C9;
+        OFS_PC_ITEMS = japanese ? 0x27DC : 0x27E6;
+    }
+    Sav1::Sav1(std::shared_ptr<u8[]> data) : Sav1(data, 0x8000) {}
```
My first thought is that this constructor shouldn't exist. Why does it currently?

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -0,0 +1,472 @@
+        }
+    }
+
+    u8 Sav1::calculateChecksum(const u8* data, size_t len)
```
Most of this logic should be extracted to a pksm::crypto function

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -0,0 +1,472 @@
+    {
+        if (playedHours() == 255 && playedMinutes() == 59 && playedSeconds() == 59)
+        {
+            data[japanese ? 0x2CA4 : 0x2CF1] = 59;  // frame count
```
I'm not sure I like these ternary operators throughout the file. Perhaps Sav1Int and Sav1JP classes? If not, perhaps most offsets should be calculated off more offset variables

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -0,0 +1,472 @@
+            data.get(), v, 0x2598, japanese ? 6 : 8, lang);
+    }
+
+    // why did they use BCD
```
BCD translation seems like it be generalized somewhere else. pksm::crypto doesn't feel right, but perhaps in (Big|Little)Endian?

SNBeast replied:
Not that I could find. You want it elsewhere?

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -0,0 +1,472 @@
+    // the PK1 and PK2 formats used by the community start with magic bytes, the second being species
+    std::unique_ptr<PKX> Sav1::pkm(u8 slot) const
+    {
+        u8 buffer[PK1Length()] = {0x01, data[partyOffset(slot)], 0xFF};
```
No variable-length arrays! Either use the largest length for the buffer or dynamically allocate it (I'd suggest the former, but might be convinced that the latter is better)

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -0,0 +1,472 @@
+            return emptyPkm();
+        }
+
+        u8 buffer[PK1Length()] = {0x01, data[boxOffset(box, slot)], 0xFF};
```
Same here with VLAs

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -0,0 +1,472 @@
+    }
+    void Sav1::pkm(const PKX& pk, u8 slot)
+    {
+        if (pk.generation() == Generation::ONE && ((pk.language() != Language::JPN) ^ (language() == Language::JPN)))
```
Second condition group should be `((pk.language() == Language::JPN) == (language() == Language::JPN))`. Also, this function should support swapping languages between JPN and INT unless you can convince me that it's completely infeasible.

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -0,0 +1,472 @@
+        }
+    }
+
+    void Sav1::trade(PKX& pk, const Date& date) const {}
```
This *probably* shouldn't be a noop

SNBeast replied:
This is, like in Gen III.

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+        }
+    }
+
+    u16 Sav2::calculateChecksum(u8* start, const u8* end)
```
A: this isn't how constness on pointers works. B: this logic should be extracted to a pksm::crypto function. Not quite certain what to call this, though: it's possible that `pksm::crypto::sum32` should be changed to a template with parameters that specify its read and return sizes

SNBeast replied:
I'm thinking `bytewiseSum16`.

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+    Language Sav2::language() const { return lang; }
+    void Sav2::language(Language v)
+    {
+        if (((lang == Language::JPN) ^ (v != Language::JPN)) &&
```
Condition should be `((lang == Language::JPN) == (v == Language::JPN)) && ((lang == Language::KOR) == (v == Language::KOR))`.

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+    // species
+    std::unique_ptr<PKX> Sav2::pkm(u8 slot) const
+    {
+        u8 buffer[PK2Length()] = {0x01, data[OFS_PARTY + 1 + slot], 0xFF};
```
No VLAs again

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+            return emptyPkm();
+        }
+
+        u8 buffer[PK2Length()] = {0x01, data[boxStart(box) + 1 + slot], 0xFF};
```
No VLAs again

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+    }
+    void Sav2::pkm(const PKX& pk, u8 slot)
+    {
+        if (pk.generation() == Generation::TWO &&
```
Same as G1 Pokémon setting: unless you can convince me it's infeasible, I want Japanese-Korean-international transfer to work. If you can't, condition should use boolean operators instead of bitwise ones again.

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+            auto pk2 = pk.clone(); // note that partyClone and clone are equivalent
+            if (applyTrade)
+            {
+                trade(*pk2);
```
Doesn't seem to handle language transfer properly

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+        }
+    }
+
+    void Sav2::trade(PKX& pk, const Date& date) const {}
```
Probably shouldn't be a noop

SNBeast replied:
It is a noop. Eggs have their met location set to where they hatch when they hatch.

piepie62 commented on source/sav/Sav2.cpp:
```diff
@@ -0,0 +1,859 @@
+    }
+
+    // needs to be fixed after GUI is fixed
+    int Sav2::maxSlot() const { return maxBoxes() * /*(japanese ? 30 : 20)*/ 30; }
```
Change that to what it should be, UI will be modified to work properly. In general for the future, just pretend the UI works properly for what you want; if I need to change these assumptions to make it easier on the UI stuff, I'll do that when I build it.

piepie62 commented on source/utils/g1text.hpp:
```diff
@@ -0,0 +1,98 @@
+{
+    char16_t tradeOTChar = u'ඞ'; // a character not expected to appear in native names, maybe it'll surprise a joker ;P
+
+    std::unordered_map<u8, char16_t> G1_EN = {{0x50, u'\0'}, {0x5D, tradeOTChar},
```
This should be a pair of constexpr arrays, not an unordered_map

piepie62 commented on source/utils/g1text.hpp:
```diff
@@ -0,0 +1,98 @@
+        {0xF7, u'1'}, {0xF8, u'2'}, {0xF9, u'3'}, {0xFA, u'4'}, {0xFB, u'5'}, {0xFC, u'6'}, {0xFD, u'7'},
+        {0xFE, u'8'}, {0xFF, u'9'}};
+
+    std::unordered_map<u8, char16_t> G1_JP = {{0x05, u'ガ'}, {0x06, u'ギ'}, {0x07, u'グ'},
```
Same here

piepie62 commented on source/utils/g2text.hpp:
```diff
@@ -0,0 +1,497 @@
+#include <unordered_map>
+#include <vector>
+
+namespace pksm::internal
```
This file and corresponding methods seem to be a mess. I'll need to discuss them with you.

piepie62 reviewed (changes requested):
Thanks for working on this! Most of it seems fine, just a few issues that'll need to get ironed out.

SNBeast committed:
Gen 1&2 Review part 1

SNBeast committed:
Gen 1&2 Review part 1.5

Actually do something I dismissed and correct big oopsies

SNBeast committed:
Gen 1&2 Review part 1.75

Fixes the language transfer to not overflow a strictly allocated Japanese PK1 or PK2

SNBeast committed:
Gen 1&2 Review Part 2

1. Replace statExperience with ev, changing ev to 16-bit, due to mistaken belief that ev needs to stay 8-bit.
2. @piepie62 improvements for G1 and G2 text handling and adding a nullptr case to getPKM.
3. Extraction of BCD handling to other places for general use (hopefully only to ever be used by Gen I).
4. Clean up handling of trade OT char, and clean up handling of incompatible-language imports.

SNBeast committed:
Fix G1&2 string writers to not overwrite terminator

SNBeast committed:
Add a failsafe for PK1&2 names when read from save

This insures that PK1&2 names have a terminator and that the designated space for terminators for a max-length string will never have a character in them (particularly ensuring that they don't have 0xFF, which may help differentiate JPN and INT in banks).

piepie62 commented on include/utils/utils.hpp:
```diff
@@ -36,27 +36,80 @@
+#include <type_traits>
 #include <vector>
 
+namespace NumberFormatUtils
```
I actually meant for this to be in the LittleEndian/BigEndian namespaces respectively. Could you move these there?

piepie62 commented on include/utils/utils.hpp:
```diff
@@ -36,27 +36,80 @@
+        return out;
+    }
+    template <typename T, size_t arrayLength = sizeof(T)>
+    constexpr std::array<u8, arrayLength> UIntegerToBigEndianBCD(T src)
```
This should have an overload that writes directly out to a u8 pointer (still with an arrayLength template parameter); look at the way that LittleEndian::convertFrom with an array works for an example of what I would prefer

piepie62 commented on source/sav/Sav1.cpp:
```diff
@@ -172,25 +172,12 @@ namespace pksm
-        BigEndian::convertFrom<u16>(
-            &data[OFS_MONEY], u16((result & 0xFFFF00) >> 8));
-        data[OFS_MONEY + 2] = u8(result & 0x0000FF);
+        std::array<u8, 3> bcd = NumberFormatUtils::UIntegerToBigEndianBCD<u32, 3>(v);
```
Use the change to this function requested to write directly instead of having a temporary

piepie62 commented on include/utils/utils.hpp:
```diff
@@ -36,27 +36,80 @@
+{
+    // Binary-Coded Decimal (BCD) conversions
+    template <typename T>
+    constexpr T BigEndianBCDtoUInteger(const u8* src, size_t arrayLength)
```
arrayLength should be a template parameter

SNBeast committed:
Gen 1&2 Review part 3

SNBeast committed:
Fixes from Crystal U and Gold U testing

piepie62 committed:
Formatting; string addition, cleanup, and fixes

piepie62 committed:
Add forgotten return

SNBeast committed:
Merge branch 'FlagBrew:master' into master

piepie62 committed:
Fix extra terminator in getString

SNBeast committed:
Fix setString1/2 repeating first character

SNBeast committed:
OT Name in Gen II is padded with 0

SNBeast committed:
Fix level being written to move 2 for PK2s

SNBeast committed:
Fix box checksum fail

SNBeast committed:
G1/2 invalidTransferReason exceptions

SNBeast committed:
Remove unneeded writes and boxCount

Removes some "false positives" for whether data was practically modified. These writes get handled by finishEditing, and a GUI that doesn't handle the fact that boxes are a gapless list instead of a 2D field would make a mess (though not noticeable in game).

SNBeast committed:
Revert data write removal in last commit for Sav2

Whoops I forgot that's necessary for eggs to be handled correctly

SNBeast committed:
Merge branch 'FlagBrew:master' into master

SNBeast committed:
Add standard* size for emulator saves with RTC

*the RTC footer is a standard, and having it as a footer is de-facto standard

Epicpkmn11 committed:
Fix nature being reset on transfer to/from PK3

Epicpkmn11 committed:
Fix PK4 -> PK3 gender, alt form, and nature

Whoops only tested PK3 -> PK4 in my first commit, PK4 -> PK3 had more issues
These were all being set incorrectly I think because the ability hadn't been set yet and that's used in PID generation, I changed it to set all of PID based stats at once to save time instead of generating three random PIDs for no reason

Octoham commented:
I'm glad this is being worked on!
That being said, I'd love to contribute to the gen 1 and 2 support. What is left to complete?

SNBeast committed:
Compilation fix

SNBeast committed:
Fix PK1/2 gender set

Epicpkmn11 committed:
Fix losing shininess on PK4 -> PK3

And some other PID changes, also simplifies anything PID-based in general
Note: This breaks compatibilty with existing code, but it's a very simple fix, just pass the shiny and TSV values to getRandomPID

SNBeast committed:
Merge pull request #1 from Epicpkmn11/fix-pk3-nature

Fix pk3 nature

SNBeast committed:
Fix PK3::hptype guards, transporter name accuracy

SNBeast commented:
Marked as mergeable.

SNBeast committed:
Fix #8

Attempt to, anyway.

SNBeast committed:
stat(Stat::HP) on Shedinja shall return 1

piepie62 committed:
Deduplicate Shedinja check
